Pico CMS for Nextcloud v1.0.0 (WIP)

PhrozenByte opened the pull request:
:construction: **WORK IN PROGRESS** :construction: 

This is, without any question, a lot... :smile: This is a combination of a total refactoring and a rewrite from scratch - but IMHO **we're very close to Pico CMS for Nextcloud version 1.0.0** :confetti_ball: :tada:

**Testing the app** is pretty straight forward:

1. [Download the tarball/ZIP](https://github.com/PhrozenByte/nextcloud-pico/archive/pico-2.0.zip) or clone my fork (see [`PhrozenByte/nextcloud-pico`](https://github.com/PhrozenByte/nextcloud-pico/tree/pico-2.0), branch `pico-2.0`) to `apps/cms_pico/` of your Nextcloud instance (requires Nextcloud 14+)
2. Run `composer install`
3. Enable the app in your Nextcloud instance
4. Check out your personal and admin settings pages of "Pico CMS" 

**PRs to my fork and comments/reviews are very welcome**! :heart: My fork also allows edits from maintainers.

This PR solves all known (at least to me... I'll reference some issues in follow-up comments) issues with the current Nextcloud app, utilizes the latest release of Pico 2.1 (currently a dev build, however, I guarantee Pico 2.1 will be released before releasing this app) and I'm furthermore willing to officially support this app (Hi everyone, I'm Pico's lead dev :v: - i.e. Pico CMS for Nextcloud will be officially supported by Pico).

Please note that many commits are rather huge and don't really address single issues, but single aspects of the app. This is rather bad for change traceability, but IMHO contributes to the final solution, since it allowed me to ignore limitations of the existing code base. Just to mention that a thousand small commits with a tons of reverts won't be much better in regards of change traceability... So... :man_shrugging: However, naturally I didn't re-invent the wheel here and changed only a few concepts, not all of them :wink: Some commits do have some elaboration on my thoughts.

@daita you really did an amazing job here! Great work! :heart: I hope you forgive me for being so boldly doing this... You invested so much time, but priorities change (or, better to say I guess, the priorities of your employer change :see_no_evil:) and I felt like this app fell a bit behind :disappointed: So many people asked me about the plans for the Nextcloud app and I just could only answer "I don't know". Since it's an amazing project and heavily contributes to Pico's visibility, I didn't want to let it fall behind... And here we are. Thanks for all of your great work @daita! :+1:

---

**Screenshots:**

<details><summary>Example website</summary>

[![Example website](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/website.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/website.png)

</details>

<details><summary>List websites</summary>

[![List websites](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/list_websites.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/list_websites.png)

</details>

<details><summary>Create website</summary>

[![Create website](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/new_website.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/new_website.png)

</details>

<details><summary>Custom themes, plugins and templates</summary>

[![Custom themes, plugins and templates](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/custom.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/custom.png)

</details>

<details><summary>Webserver config</summary>

[![Webserver config](https://raw.githubusercontent.com/nextcloud/cms_pico/dc61780f23f091fec629c62e332eace4f6807875/screenshots/apache.png)](https://github.com/nextcloud/cms_pico/blob/dc61780f23f091fec629c62e332eace4f6807875/screenshots/apache.png)

</details>

---

**Some thoughts about some particular features:**

- Pico CMS for Nextcloud now supports both `mod_proxy`-based URLs (like `https://example.com/sites/my_pico_site`) and generic URLs (like `https://example.com/index.php/apps/cms_pico/pico/my_pico_site`). Currently the latter is shown by default (since it works out-of-the-box and webserver configuration was the most often asked question and source of problems until now) and the shorter form is just mentioned. I'll add a simple config option to enable short URLs after the admin has configured the webserver appropriately. - *Done* :+1:

- I'm planning to add a "Add custom plugin" feature just like for themes, allowing admins to add plugins to Pico. Plugins are tested for compatibility (Pico will throw an exception otherwise). We might even extend this to enabling plugins on a per-website basis. However, the latter is still a long way off. - *Done* :+1:

- Handling assets (content assets, theme assets and plugin assets) still isn't optimal (especially regarding browser caching). To solve the remaining issues I'll add a feature to Pico to support `%assets_url%` and `%plugins_url%` next to the already existing `%theme_url%` (+ corresponding Twig variables; actually this is already implemented in Pico 2.1). These variables can then be used by Pico CMS for Nextcloud to create version-specific URLs utilizing Nextcloud's ETag (content assets) resp. some random string that is generated when a theme/plugin is published or updated. Not sure about the details yet. - *Done* :+1:

- The two features listed above are currently just available for Pico plugins. I'll enable the same functionality for Pico themes, too. - *Done* :+1:

- I'm no big fan of unit tests (not because I don't like them, I usually just don't have time writing them...). All existing tests are still working, but I didn't write any new.

- I don't really know how Transifex works :confused: `de.js`, `de.json`, `de_DE.js`, `de_DE.json`, `en_GB.js` and `en_GB.json` are all fully up-to-date, but how can we import them to Transifex?

- I've absolutely no idea how this whole release process works (`.drone.yml` :question:)...

karlitschek commented:
@PhrozenByte Wow. This is awesome. Thanks a lot for taking this over. This has the potential to become one of the most useful Nextcloud apps!!!!!

karlitschek commented:
@PhrozenByte by the way. Are you open for some UI feedback from the Nextcloud UI team? There might be some things to improve ;-)

PhrozenByte commented:
Thanks @karlitschek! Yeah, definitely, feedback is very welcome! :+1: 

I was trying to resemble some of Nextcloud's UI elements. Sadly I couldn't really find much documentation or, what would have been even better, re-usable components. I only found rather advanced stuff like Nextcloud's popover menu, but nothing about the basic stuff like tables or forms. The only thing I could find was [`.app-content-list` for sidebars](https://docs.nextcloud.com/server/16/developer_manual/design/list.html) - but this is definitely not intended to be used the way I did :smile: So, yeah, UI feedback is definitely welcome, too.

karlitschek commented:
@PhrozenByte very cool!

@nextcloud/designers @jancborchardt Maybe you can help here a bit? :-)

MorrisJobke commented:
Hi,

let me answer one of your questions:

> * I don't really know how Transifex works 😕 `de.js`, `de.json`, `de_DE.js`, `de_DE.json`, `en_GB.js` and `en_GB.json` are all fully up-to-date, but how can we import them to Transifex?

We extract source strings from the app and push them to transifex. Then we fetch updated translations from transifex and convert them into our format. Thus changes to translations will be overwritten by this sync and should be made on transifex once this PR is merged and thus has the updated source strings. If you really really want to import them, then we need to write a script that converts those to a PO file and then we can do a one time transifex import, but this will forcefully overwrite all strings on transifex and should be handled with care. I would prefer to manually go over them on the transifex side (there are also people that might help you with that)

PhrozenByte commented:
@MorrisJobke Thanks for the response! Hmm... Not sure which way to go. It's important to note that there are just very few strings unchanged. I guess I'll leave this up to you guys, I'll keep the German translation up-to-date anyway.

jancborchardt commented:
@PhrozenByte hey there! :) Haven’t tried out the app yet, will check out your fork! :)

Also, are you coming to the Nextcloud Conference in Berlin? https://nextcloud.com/conf – would be really cool to have you, and you could give a lightning talk on the Pico CMS app!

jancborchardt commented:
Getting an error on trying to enable the app:

Browser console:
```
Content Security Policy: The page’s settings blocked the loading of a resource at eval (“script-src”).
Object { apps: Getter & Setter, categories: Getter & Setter, updateCount: Getter & Setter, loading: Getter & Setter, loadingList: Getter & Setter, … }
 
Object { appId: "cms_pico", error: Error }
apps.js:38:10
Object { apps: Getter & Setter, categories: Getter & Setter, updateCount: Getter & Setter, loading: Getter & Setter, loadingList: Getter & Setter, … }
 
{…}
​
appId: "cms_pico"
​
error: Error
​​
columnNumber: 71354
​​
config: Object { url: "/nextcloud/index.php/settings/apps/enable", method: "post", data: "{\"appIds\":[\"cms_pico\"],\"groups\":[]}", … }
​​
fileName: "http://localhost/nextcloud/settings/js/vue-settings-apps-users-management.js"
​​
isAxiosError: true
​​
lineNumber: 7
​​
message: "Request failed with status code 500"
​​
request: XMLHttpRequest { readyState: 4, timeout: 0, withCredentials: false, … }
​​
response: Object { data: {…}, status: 500, statusText: "Internal Server Error", … }
​​
stack: "t.exports@http://localhost/nextcloud/settings/js/vue-settings-apps-users-management.js:7:71354\nt.exports@http://localhost/nextcloud/settings/js/vue-settings-apps-users-management.js:45:2565\nt.exports/</l.onreadystatechange@http://localhost/nextcloud/settings/js/vue-settings-apps-users-management.js:7:70293\n"
​​
toJSON: function toJSON()​​
<prototype>: {…}
​​​
constructor: function Error()
​​​
message: ""
​​​
name: "Error"
​​​
stack: "t.exports@http://localhost/nextcloud/settings/js/vue-settings-apps-users-management.js:7:71354\nt.exports@http://localhost/nextcloud/settings/js/vue-settings-apps-users-management.js:45:2565\nt.exports/</l.onreadystatechange@http://localhost/nextcloud/settings/js/vue-settings-apps-users-management.js:7:70293\n"
​​​
toSource: function toSource()
​​​
toString: function toString()
​​​
<get stack()>: function stack()
​​​
<set stack()>: function stack()
​​​
<prototype>: {…
```


Server log:
```
{"reqId":"UWcZu0h83k8HNp25aIbL","level":0,"time":"2019-07-23T15:27:05+00:00","remoteAddr":"127.0.0.1","user":"jan","app":"cms_pico","method":"POST","url":"\/nextcloud\/index.php\/settings\/apps\/enable","message":"Copying Pico CMS config \u2026","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"17.0.0.1"}
{"reqId":"UWcZu0h83k8HNp25aIbL","level":2,"time":"2019-07-23T15:27:05+00:00","remoteAddr":"127.0.0.1","user":"jan","app":"cms_pico","method":"POST","url":"\/nextcloud\/index.php\/settings\/apps\/enable","message":"Replacing config file \"config.yml.template\"","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"17.0.0.1"}
{"reqId":"UWcZu0h83k8HNp25aIbL","level":0,"time":"2019-07-23T15:27:05+00:00","remoteAddr":"127.0.0.1","user":"jan","app":"cms_pico","method":"POST","url":"\/nextcloud\/index.php\/settings\/apps\/enable","message":"Copying Pico CMS templates \u2026","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"17.0.0.1"}
{"reqId":"UWcZu0h83k8HNp25aIbL","level":2,"time":"2019-07-23T15:27:05+00:00","remoteAddr":"127.0.0.1","user":"jan","app":"cms_pico","method":"POST","url":"\/nextcloud\/index.php\/settings\/apps\/enable","message":"Replacing template \"empty\"","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"17.0.0.1"}
{"reqId":"UWcZu0h83k8HNp25aIbL","level":2,"time":"2019-07-23T15:27:05+00:00","remoteAddr":"127.0.0.1","user":"jan","app":"cms_pico","method":"POST","url":"\/nextcloud\/index.php\/settings\/apps\/enable","message":"Replacing template \"sample_pico\"","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"17.0.0.1"}
{"reqId":"UWcZu0h83k8HNp25aIbL","level":0,"time":"2019-07-23T15:27:06+00:00","remoteAddr":"127.0.0.1","user":"jan","app":"cms_pico","method":"POST","url":"\/nextcloud\/index.php\/settings\/apps\/enable","message":"Publishing Pico CMS themes \u2026","userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"17.0.0.1"}
{"reqId":"UWcZu0h83k8HNp25aIbL","level":3,"time":"2019-07-23T15:27:06+00:00","remoteAddr":"127.0.0.1","user":"jan","app":"settings","method":"POST","url":"\/nextcloud\/index.php\/settings\/apps\/enable","message":{"Exception":"OCP\\Files\\NotPermittedException","Message":"","Code":0,"Trace":[{"file":"\/home\/jan\/nextcloud\/apps\/cms_pico\/lib\/Files\/AbstractNode.php","line":80,"function":"delete","class":"OCA\\CMSPico\\Files\\AbstractLocalNode","type":"->","args":[]},{"file":"\/home\/jan\/nextcloud\/apps\/cms_pico\/lib\/Migration\/AppDataRepairStep.php","line":163,"function":"empty","class":"OCA\\CMSPico\\Files\\AbstractNode","type":"->","args":[]},{"file":"\/home\/jan\/nextcloud\/apps\/cms_pico\/lib\/Migration\/AppDataRepairStep.php","line":99,"function":"publishThemes","class":"OCA\\CMSPico\\Migration\\AppDataRepairStep","type":"->","args":[]},{"file":"\/home\/jan\/nextcloud\/lib\/private\/Repair.php","line":98,"function":"run","class":"OCA\\CMSPico\\Migration\\AppDataRepairStep","type":"->","args":[{"__class__":"OC\\Repair"}]},{"file":"\/home\/jan\/nextcloud\/lib\/private\/legacy\/app.php","line":974,"function":"run","class":"OC\\Repair","type":"->","args":[]},{"file":"\/home\/jan\/nextcloud\/lib\/private\/Installer.php","line":158,"function":"executeRepairSteps","class":"OC_App","type":"::","args":["cms_pico",{"step":"OCA\\CMSPico\\Migration\\AppDataRepairStep"}]},{"file":"\/home\/jan\/nextcloud\/settings\/Controller\/AppSettingsController.php","line":447,"function":"installApp","class":"OC\\Installer","type":"->","args":["cms_pico"]},{"file":"\/home\/jan\/nextcloud\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":166,"function":"enableApps","class":"OC\\Settings\\Controller\\AppSettingsController","type":"->","args":[["cms_pico"],[]]},{"file":"\/home\/jan\/nextcloud\/lib\/private\/AppFramework\/Http\/Dispatcher.php","line":99,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OC\\Settings\\Controller\\AppSettingsController"},"enableApps"]},{"file":"\/home\/jan\/nextcloud\/lib\/private\/AppFramework\/App.php","line":126,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->","args":[{"__class__":"OC\\Settings\\Controller\\AppSettingsController"},"enableApps"]},{"file":"\/home\/jan\/nextcloud\/lib\/private\/AppFramework\/Routing\/RouteActionHandler.php","line":47,"function":"main","class":"OC\\AppFramework\\App","type":"::","args":["OC\\Settings\\Controller\\AppSettingsController","enableApps",{"__class__":"OC\\AppFramework\\DependencyInjection\\DIContainer"},{"_route":"settings.AppSettings.enableApps"}]},{"function":"__invoke","class":"OC\\AppFramework\\Routing\\RouteActionHandler","type":"->","args":[{"_route":"settings.AppSettings.enableApps"}]},{"file":"\/home\/jan\/nextcloud\/lib\/private\/Route\/Router.php","line":297,"function":"call_user_func","args":[{"__class__":"OC\\AppFramework\\Routing\\RouteActionHandler"},{"_route":"settings.AppSettings.enableApps"}]},{"file":"\/home\/jan\/nextcloud\/lib\/base.php","line":975,"function":"match","class":"OC\\Route\\Router","type":"->","args":["\/settings\/apps\/enable"]},{"file":"\/home\/jan\/nextcloud\/index.php","line":42,"function":"handleRequest","class":"OC","type":"::","args":[]}],"File":"\/home\/jan\/nextcloud\/apps\/cms_pico\/lib\/Files\/AbstractLocalNode.php","Line":125,"CustomMessage":"--"},"userAgent":"Mozilla\/5.0 (X11; Ubuntu; Linux x86_64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"17.0.0.1"}
```

What am I missing? I did `composer install` and it ran through without issues.

PhrozenByte commented:
A `NotPermittedException` is being thrown when trying to empty the `appdata_public/themes/` folder. I guess that's due to file permissions being wrong, which should only occur when being installed separately and not via the app store (what isn't possible right now though). Make sure the file owner, group and permissions of all of the app's files are identical to those of Nextcloud's files. However, I haven't tested the latest changes (last tested on e2e60a1) yet (I'll catch up on that later), so there might be another issue, but please check the file permissions first.

Not sure about whether I should include some code for trying to fix the file permissions automatically... What do you guys think?

daita commented:
> Not sure about whether I should include some code for trying to fix the file permissions automatically... What do you guys think?

Just complement the README file with additional prerequisite when using repo.

juliushaertl commented:
> I was trying to resemble some of Nextcloud's UI elements. Sadly I couldn't really find much documentation or, what would have been even better, re-usable components. I only found rather advanced stuff like Nextcloud's popover menu, but nothing about the basic stuff like tables or forms. The only thing I could find was .app-content-list for sidebars - but this is definitely not intended to be used the way I did smile So, yeah, UI feedback is definitely welcome, too.

We are currently building a Vue.js based component library for UI elements that can be used across apps. I'm also working on a set of settings elements for this (see https://github.com/nextcloud/nextcloud-vue/issues/498 for details), but apart from that I'd suggest to have a look at the structure of the settings sections of core apps for now.

jancborchardt commented:
> I guess that's due to file permissions being wrong, which should only occur when being installed separately and not via the app store (what isn't possible right now though).

Yeah, I installed it via git.

> Not sure about whether I should include some code for trying to fix the file permissions automatically... What do you guys think?

Would be good because even though I’m fairly technical (normally a designer ;), needing to fix permissions is where I’d normally tap out.


The permissions of my main Nextcloud folder:
```
[…]
drwxr-xr-x  33 jan      jan        4096 Jun  7 17:25  3rdparty
-rw-rw-r--   1 jan      jan         745 Jun 20 18:43  Makefile
-rw-rw-r--   1 jan      jan         103 Jun 20 18:43  .babelrc
-rw-rw-r--   1 jan      jan        2287 Jun 20 18:43  status.php
drwxrwxr-x 565 jan      jan       20480 Jun 28 11:54  node_modules
drwxr-xr-x   3 jan      jan        4096 Jul 15 17:12  .github
drwxrwxr-x   2 jan      www-data   4096 Jul 15 17:12  config
drwxr-xr-x  15 jan      jan        4096 Jul 15 17:12  settings
-rw-rw-r--   1 jan      jan        2115 Jul 15 17:12  webpack.common.js
drwxrwxr-x  68 jan      www-data   4096 Jul 23 16:33  apps
drwxr-xr-x   9 jan      jan        4096 Jul 23 17:28  build
drwxr-xr-x  24 jan      jan        4096 Jul 23 17:28  core
drwxr-xr-x   6 jan      jan        4096 Jul 23 17:28  lib
-rw-rw-r--   1 jan      jan      356930 Jul 23 17:28  package-lock.json
-rw-rw-r--   1 jan      jan        2766 Jul 23 17:28  package.json
drwxr-xr-x  23 jan      jan        4096 Jul 23 17:28  .
drwxr-xr-x  10 jan      jan        4096 Jul 23 17:30  .git
drwxr-xr-x 100 jan      jan       49152 Jul 23 18:27  ..
drwxrwx---  15 www-data www-data   4096 Jul 23 20:02  data
```

Of cms_pico:
```
jan@Rechenknecht:~/nextcloud/apps/cms_pico$ ls -la
total 140
drwxrwxr-x 68 jan www-data  4096 Jul 23 16:33 ..
drwxrwxr-x  2 jan jan       4096 Jul 23 16:33 .tx
-rw-rw-r--  1 jan jan      32386 Jul 23 16:33 LICENSE
-rw-rw-r--  1 jan jan       6762 Jul 23 16:33 .drone.yml
-rw-rw-r--  1 jan jan        298 Jul 23 16:33 CHANGELOG.md
-rw-rw-r--  1 jan jan       1787 Jul 23 16:33 Makefile
drwxrwxr-x  2 jan jan       4096 Jul 23 16:33 img
drwxrwxr-x  8 jan jan       4096 Jul 23 17:14 vendor
drwxrwxr-x  3 jan jan       4096 Jul 23 17:29 appdata
-rw-rw-r--  1 jan jan       3745 Jul 23 17:29 .scrutinizer.yml
-rw-rw-r--  1 jan jan         19 Jul 23 17:29 .gitignore
drwxrwxr-x  6 jan jan       4096 Jul 23 17:29 Pico
-rw-rw-r--  1 jan jan       1181 Jul 23 17:29 README.md
drwxrwxr-x  2 jan jan       4096 Jul 23 17:29 css
-rw-rw-r--  1 jan jan      11484 Jul 23 17:29 composer.lock
-rw-rw-r--  1 jan jan        319 Jul 23 17:29 composer.json
drwxrwxr-x  2 jan jan       4096 Jul 23 17:29 appinfo
drwxrwxr-x 16 jan jan       4096 Jul 23 17:29 .
drwxrwxr-x  3 jan jan       4096 Jul 23 17:29 js
drwxrwxr-x  2 jan jan       4096 Jul 23 17:29 l10n
drwxrwxr-x 11 jan jan       4096 Jul 23 17:29 lib
drwxrwxr-x  3 jan jan       4096 Jul 23 17:29 tests
drwxrwxr-x  2 jan jan       4096 Jul 23 17:29 templates
drwxrwxr-x  2 jan jan       4096 Jul 23 17:29 screenshots
drwxrwxr-x  8 jan jan       4096 Jul 23 17:29 .git
```

PhrozenByte commented:
@juliushaertl Hmm okay... Not sure what I'm supposed to do now :confused: The app's admin settings page consist of multiple lists of elements (themes, plugins and templates) with two actions (see the "Custom themes and templates" screenshot above; I tried to resemble `.app-content-list` here). The app's personal settings page consists of a list of websites with various meta data (see the "List websites" screenshot above; I tried to resemble the main table of the `files` app here) and a simple form with various input fields (see the "Create website" screenshot above; AFAIK there's nothing similar in the core settings yet). Not sure whether my current design comes close to fairly resembling Nextcloud's UI elements, I'm no designer :smile:

@daita Looks like Nextcloud server circumvents this by not having the `data` folder commited and rather creating the folder on install. Naturally install will still fail if Nextcloud couldn't create the folder, but I was just thinking about creating the `appdata_public` folder in the `Version10000` migration step, so that we can throw a nice exception when enabling the app. We could even send a notification. Might be a more user-friendly solution. What do you think? Will update the `README.md` anyways of course :smile: 

@jancborchardt Hmm... That's weird. Your `cms_pico` folder is that of the current stable version, not the version of this PR. Since your logs prove you're running the latest version, I suppose you accidently just pasted the directory listing of the wrong folder? Anyways, it looks like your webserver is running as user `www-data`. Just change the owner and group of the app's `appdata_public` folder to `www-data` using this command (from within the right folder of course): `chown www-data:www-data -R appdata_public`

juliushaertl commented:
> Hmm okay... Not sure what I'm supposed to do now confused The app's admin settings page consist of multiple lists of elements (themes, plugins and templates) with two actions (see the "Custom themes and templates" screenshot above; I tried to resemble .app-content-list here). The app's personal settings page consists of a list of websites with various meta data (see the "List websites" screenshot above; I tried to resemble the main table of the files app here) and a simple form with various input fields (see the "Create website" screenshot above; AFAIK there's nothing similar in the core settings yet). Not sure whether my current design comes close to fairly resembling Nextcloud's UI elements, I'm no designer smile

The screenshots look pretty decent to me, but maybe @jancborchardt can do some nitpicking on the settings layout :wink: 

> Hmm... That's weird. Your cms_pico folder is that of the current stable version, not the version of this PR. Since your logs prove you're running the latest version, I suppose you accidently just pasted the directory listing of the wrong folder? Anyways, it looks like your webserver is running as user www-data. Just change the owner and group of the app's appdata_public folder to www-data using this command (from within the right folder of course): chown www-data:www-data -R appdata_public

The issue with the approach of putting additional files into the app folder itself is that you cannot be sure it is writable on all setups. Can you provide some more details what data is written into that directory? Is it uploaded by the user or the admin? Usually app data uploaded afterwards by apps should be stored in either the user root or the appdata folder.

PhrozenByte commented:
> The issue with the approach of putting additional files into the app folder itself is that you cannot be sure it is writable on all setups. Can you provide some more details what data is written into that directory? Is it uploaded by the user or the admin? Usually app data uploaded afterwards by apps should be stored in either the user root or the appdata folder.

Custom themes and plugins must be accessible via webbrowser, otherwise it would be impossible to use them (just think of theme assets, like a `style.css`). The `appdata` folder is a storage folder that isn't accessible via webbrowser for obvious reasons. There are two solutions to this issue: Either route all traffic through PHP and Nextcloud to serve these files (something I actually implemented first, see a9f532a), however, this solution has an absolutely horrible impact on performance. Another solution is to copy a theme's and plugin's files to a place where the webbrowser can access them. That's the `appdata_public` folder. However, we can't just tell the admin to upload files directly to this folder, since it will be reset when updating the app. Thus we tell the admin to upload custom plugins and themes to Nextcloud's `appdata` folder and let the app copy the files to the app's `appdata_public` folder.

MorrisJobke commented:
> Custom themes and plugins must be accessible via webbrowser, otherwise it would be impossible to use them (just think of theme assets, like a `style.css`). The `appdata` folder is a storage folder that isn't accessible via webbrowser for obvious reasons. There are two solutions to this issue: Either route all traffic through PHP and Nextcloud to serve these files (something I actually implemented first, see [a9f532a](https://github.com/nextcloud/cms_pico/commit/a9f532a0ff730065209d44f9c56366b5857718e0)), however, this solution has an absolutely horrible impact on performance. Another solution is to copy a theme's and plugin's files to a place where the webbrowser can access them. That's the `appdata_public` folder. However, we can't just tell the admin to upload files directly to this folder, since it will be reset when updating the app. Thus we tell the admin to upload custom plugins and themes to Nextcloud's `appdata` folder and let the app copy the files to the app's `appdata_public` folder.

Just to give you some background of why there is a restriction: The reason that this is not served by the webserver is that it would allow code injections by users into the Nextcloud itself. Best would be if that content is served under a different domain to avoid security issues here.

jancborchardt commented:
> Just to give you some background of why there is a restriction: The reason that this is not served by the webserver is that it would allow code injections by users into the Nextcloud itself. Best would be if that content is served under a different domain to avoid security issues here.

@MorrisJobke That of course would make the simplicity of it way harder again. :) I can understand it though – is there other ways we can solve this? Like
- Let people know about this problem in the app description
- Or does this happen anyway with static content (HTML and CSS not, right? But the JS part?)?
- Or would the app need to sanitize stuff before serving?
- … anything else?

PhrozenByte commented:
@jancborchardt Pico CMS for Nextcloud sanitizes contents before serving them using HTMLPurifier. Only admins with access to the FTP server (i.e. Nextcloud's files) can add custom themes and therefore use JavaScript. So this shouldn't be a issue here :smiley: 

About the directory creation issue: Honestly I don't really see a problem, I'll simply add a nice error message telling users that the webserver wasn't able to create a required directory. Just like what Nextcloud does when Nextcloud is being installed. The whole issue doesn't affect app store users, so this won't happen too often anyway.

MorrisJobke commented:
> @MorrisJobke That of course would make the simplicity of it way harder again. :) I can understand it though – is there other ways we can solve this?

Sanitizing is always possible - but then needs to be served via PHP and not by the Webserver directly. That is the same reason why files are always downloaded and some images (like SVGs) are only shown as either preview or with the attachment header (which then forces browsers to download instead of inline it). The problem is, that in this way we eliminate a whole category of attack vectors (user content & stored XSS), because we simply don't serve user content (especially JS) under the same domain. Because then this served JS could execute in the context of our own code and do attacks on users without any chance from us to protect against this. That is the reason, why GitHub for example serves raw files from githubusercontent.com. In that way it can't be executed on GitHub.com even if it is valid JS and could somehow be injected in the website.

PhrozenByte commented:
It has been a while... But I assume the code to be feature complete for v1.0.0, no open code-related ToDos (besides releasing Pico 2.1, but that's another topic).

The only code-related nice-to-have feature are more and updated unit tests. I don't assume this to be a release-blocker. **Nevertheless, help is highly appreciated!** :+1: 

Before merging there are some more, non-code related ToDos:

- [ ] Write a new `README.md` and re-use its contents for `appinfo/info.xml` (i.e. the App Store) and the main description in `templates/settings.personal.php`. Also create new screenshots with some sample contents (and with an English UI). I hope I'll find some time the week after next

- [ ] Import `l10n/{de,de_DE,en_GB}.{json,js}` to Transifex. No idea how to do this... @MorrisJobke: I guess a fresh import is better, there are virtually no overlapping strings besides e.g. "Name". How can we do this?

- [ ] Make the release process v1.0.0-ready (`.drone.yml` seems to reference Nextcloud 12, but the app requires Nextcloud v14 - v17); @daita: I definitely need some help here

- [ ] Do some more testing... The only thing I've tested is a fresh install on Nextcloud 16. I'll do some tests with Nextcloud 14, 15 and 17 as well as updating old installations (I did a single upgrade some time ago, worked pretty well, thus I'm pretty sure that I've considered all changes in the `Version010000` migration)

**Code reviews are highly appreciated!** :heart:
(I'm pretty sure @daita wants to do a code review :smile:)

I'm currently aiming for a release within the next 3 weeks (i.e. within August). However, next week I'll have basically no time for Pico...

PhrozenByte commented:
Here we go. I've asked for feedback and testers on [help.nextcloud.com](https://help.nextcloud.com/t/beta-needs-testing-pico-cms-for-nextcloud-v1-0-0/58718) and created a [`v1.0.0-beta.1`](https://github.com/nextcloud/cms_pico/releases/tag/v1.0.0-beta.1) release. Additionally I skimmed through old Issues here on GitHub and closed a lot of them. I might implement one or the another before release, we'll see.

Open issues:

- Build & Release process
- Transifex
- Scrutinizer

**Looking forward to any more feedback and/or code reviews!** :heart:

juliushaertl commented:
>  Import l10n/{de,de_DE,en_GB}.{json,js} to Transifex. No idea how to do this... @MorrisJobke: I guess a fresh import is better, there are virtually no overlapping strings besides e.g. "Name". How can we do this?

While ther probably might be a way to script this and import it though the transifex api i would still say that the effort bigger than just waiting for the first sync of transifex and then manually adding the translations in transifex again.

> Scrutinizer needs to be updated (.scurinizer.yml grey_question). Many issues related to code styling there aren't really issues (I mostly follow PHP-FIG's coding standards)... Not sure how to fix this either.

Te .scrutinizer.yml seems to be quite old, I would try to just replace it with a rather default one for PHP https://scrutinizer-ci.com/docs/tools/php/php-scrutinizer/#configuring-php-scrutinizer

Since I've seen you are located in Germany, maybe you want to joint our Nextcloud conference in Berlin September 14-15 https://nextcloud.com/conf-2019 Would be cool to have you there, maybe you even want to give a short lightning talk about the work you have done with Pico CMS? Contributors get a free ticket btw. :wink:

PhrozenByte commented:
> While ther probably might be a way to script this and import it though the transifex api i would still say that the effort bigger than just waiting for the first sync of transifex and then manually adding the translations in transifex again.

Okay, so nothing to do at the moment. Got it, thanks @juliushaertl :+1: 

> Te .scrutinizer.yml seems to be quite old, I would try to just replace it with a rather default one for PHP https://scrutinizer-ci.com/docs/tools/php/php-scrutinizer/#configuring-php-scrutinizer

Looked into Scrutinizer a bit more deeply. Wow, one would assume that these guys document their services well, but no, the docs are incredibly bad. No structure at all... I really have absolutely no idea what I'm doing there, rather trial & error. In the meantime I even broke it beyond repair; something about an invalid config, even after restoring the original one. No idea what was going on there... I stripped down the config and added a `.phpcs.xml` for PHP_CodeSniffer. I know PHP_CodeSniffer at least. And it allows me to run `phpcs` locally, too. No need for these Scrutinizer coding style checks...

I've fixed all Scrutinizer issues, however, no idea whether Scrutinizer is actually doing what it is supposed to do or just skips a lot of (sort of important) checks now... Just let me know if something is wrong there... :smile: 

> Since I've seen you are located in Germany, maybe you want to joint our Nextcloud conference in Berlin September 14-15 https://nextcloud.com/conf-2019 Would be cool to have you there, maybe you even want to give a short lightning talk about the work you have done with Pico CMS? Contributors get a free ticket btw. :wink:

@daita already asked me whether I want to attend. The conference isn't far away from my working place. However, unfortunately I'm not in Berlin this weekend, business appointments :sob: Too bad. Hopefully next year :smiley:

juliushaertl commented:
> I've fixed all (meaningful) Scrutinizer issues, however, no idea whether Scrutinizer is actually doing what it is supposed to do or just skips a lot of (sort of important) checks now... Just let me know if something is wrong there... smile

All fine. This is totally up to the repo, I moved some of my apps to codacy since I also wasn't to happy with Scrutinizer.

> Already asked me whether I want to attend. The conference isn't far away from my working place. However, unfortunately I'm not in Berlin this weekend, business appointments sob Too bad. Hopefully next year smiley

Ah, ok. Hopefully next year then. We also have regular contributor weeks in Stuttgart in case this might be nearer :wink:

daita commented:
@rullzer @morrisjobke I shared the keys to @phrozenByte. Can you add his account as maintainer on the appstore?

@phrozenByte can you give us your account on apps.nc.c?

PhrozenByte commented:
> @rullzer @MorrisJobke I shared the keys to @PhrozenByte. Can you add his account as maintainer on the appstore?
> 
> @PhrozenByte can you give us your account on apps.nc.c?

Thanks @daita :+1: My apps.nextcloud.com account is `PhrozenByte`

acs-ferreira commented:
Great work! Is there an ETA at this stage?

PhrozenByte commented:
> > @rullzer @MorrisJobke I shared the keys to @PhrozenByte. Can you add his account as maintainer on the appstore?
> > @PhrozenByte can you give us your account on apps.nc.c?
> 
> Thanks @daita :+1: My apps.nextcloud.com account is `PhrozenByte`

@rullzer @MorrisJobke did you enable my account (it's `PhrozenByte`) on apps.nextcloud.com for uploading new releases of `cms_pico`? Not sure whether there's a special "My apps" list on apps.nextcloud.com, however, https://apps.nextcloud.com/account/transfer-apps (where I'd expect the app to show up) lists nothing for me.

I'd like to release a beta 2 on the app store. The app itself is feature complete, however, I'd like to give Pico 2.1 some more testing before releasing Pico 2.1 as final. Since I don't want to make the app final until Pico 2.1 is final, there's going to be a beta 2 - just with a larger user base. I expect the app to be stable nevertheless.

PhrozenByte commented:
**The second beta of [Pico CMS for Nextcloud](https://apps.nextcloud.com/apps/cms_pico) is out now!** :confetti_ball: :tada:

You can download Pico CMS for Nextcloud's release archive from here: [`cms_pico-1.0.0-beta.2.tar.gz`](https://github.com/nextcloud/cms_pico/releases/download/v1.0.0-beta.2/cms_pico-v1.0.0-beta.2.tar.gz) 

**Heads up!** If you're currently running Pico CMS for Nextcloud `v1.0.0-beta.1` you **MUST** check the *"I'm running `v1.0.0-beta.1` right now"* section in the announcement before upgrading. This is also true when upgrading to `v1.0.0` later - **you'll break your installation otherwise!**

---

**Full Announcement**

:arrow_right: [[Beta] [Needs Testing] Pico CMS for Nextcloud v1.0.0](https://help.nextcloud.com/t/beta-needs-testing-pico-cms-for-nextcloud-v1-0-0/58718/16) on Nextcloud Help :arrow_left:

PhrozenByte commented:
**The final release of Pico CMS for Nextcloud v1.0.0 is out now!** :confetti_ball: :tada:

You can download Pico CMS for Nextcloud's release archive from here: [`cms_pico-1.0.0.tar.gz`](https://github.com/nextcloud/cms_pico/releases/download/v1.0.0/cms_pico-v1.0.0.tar.gz)

**Heads up!** If you're currently running Pico CMS for Nextcloud `v1.0.0-beta.1` you **MUST** upgrade to `v1.0.0-beta.2` first before upgrading to `v1.0.0` later - **you'll break your installation otherwise!** Please refer to the [announcement of `v1.0.0-beta.2`](https://help.nextcloud.com/t/beta-needs-testing-pico-cms-for-nextcloud-v1-0-0/58718/16) for more details.

---

**Full Announcement**

:arrow_right: [[Beta] [Needs Testing] Pico CMS for Nextcloud v1.0.0](https://help.nextcloud.com/t/beta-needs-testing-pico-cms-for-nextcloud-v1-0-0/58718/21) on Nextcloud Help :arrow_left:

PhrozenByte commented:
Just for reference: I just "imported" the development translation files (`l10n/de.json`, `l10n/de_DE.json` and `l10n/en_GB.json`) into [Transifex](https://www.transifex.com/nextcloud/nextcloud/cms_pico/) using the following "script" on the command line (super dirty, but does the job):

```
$ php -r '$l10n = json_decode(file_get_contents("l10n/de.json"), true)["translations"]; foreach ($l10n as $msgid => $msgstr) { echo "msgid \"" . str_replace("\"", "\\\"", $msgid) . "\"\n"; echo "msgstr \"" . str_replace("\"", "\\\"", $msgstr) . "\"\n\n"; }'
```

One just has to download the "Download file to translate" on Transifex and replace all language strings with the output of the above command. Upload the resulting `.po` file using the "Upload file" on Transifex and you're ready to go.

PhrozenByte commented:
> > * I don't really know how Transifex works confused `de.js`, `de.json`, `de_DE.js`, `de_DE.json`, `en_GB.js` and `en_GB.json` are all fully up-to-date, but how can we import them to Transifex?
> 
> We extract source strings from the app and push them to transifex. Then we fetch updated translations from transifex and convert them into our format. Thus changes to translations will be overwritten by this sync and should be made on transifex once this PR is merged and thus has the updated source strings. If you really really want to import them, then we need to write a script that converts those to a PO file and then we can do a one time transifex import, but this will forcefully overwrite all strings on transifex and should be handled with care. I would prefer to manually go over them on the transifex side (there are also people that might help you with that)

@MorrisJobke Looks like Transifex wasn't able to import all l10n strings, a few are missing (because the untranslated string are stored in the appconfig and passed to `IL10N` later). How can I add some strings to Trasifex manually, or is there a way/syntax how I can explicitly tell Transifex that a arbitrary string is indeed a l10n string?

Here's an example:

https://github.com/nextcloud/cms_pico/blob/5b9a40030c6731f68f01e4eae07913e6cbfdc4a6/lib/Model/Plugin.php#L152-L156

I was thinking about some sort of comment syntax that tells Transifex that there's an l10n string, maybe something like this

```php
				// L10N: Incompatible plugin: Plugin class "{class}" not found.
				throw new PluginNotCompatibleException( 
					$this->getName(), 
					'Incompatible plugin: Plugin class "{class}" not found.', 
					[ 'class' => $className ] 
				);
``` 

How can I achieve this?

PhrozenByte commented:
@MorrisJobke 

I've got another issue with Transifex: Looks like I don't have elevated rights for the `cms_pico` project (my username there is PhrozenByte, too). I can't change l10n strings, only draft suggestions. I think I can't review l10n strings either. How can I get those elevated rights?

Ludovicis commented:
Hello PhrozenByte,

It's OK also for the French language in transifex (But no reviewed, i don't know who is the examiner...).
Thanks again for your work

juliushaertl commented:
@jospoortvliet probably wants to do some marketing about the new app release :wink:

jospoortvliet commented:
Yes! We should talk about this! @Ludovicis please please email me (my first name @ the nextcloud.com server) - we should blog. Pretty please because this is awesome :rocket:

jospoortvliet commented:
@PhrozenByte 👆👆👆

PhrozenByte commented:
@jospoortvliet Sorry, missed your edit. Just sent you an email. Looking forward to it, thanks! :+1:
